asmcomp/cmmgen.ml

```diff
@@ -300,12 +300,50 @@ let invert_then_else = function
   | Then_false_else_true -> Then_true_else_false
   | Unknown -> Unknown
 
+let rec simplify_cond = function
+  | Cop(Ccmpi cmpop
+       , [ Cop(Caddi,[Cop(Clsl,[left;Cconst_int(1,_)],_);Cconst_int(1,_)],_)
+         ; Cconst_int (right_i,right_dbg)
+         ]
+       , dbg_cmp
+       ) ->
+      Cop (Ccmpi cmpop, [ left; Cconst_int (right_i asr 1, right_dbg)], dbg_cmp)
```

Comment:
This optimisation isn't quite right. The optimisation rewrites:

    compare ((left lsl 1) + 1) right
       =>
    compare left (right asr 1)

I think this is invalid unless `right` is odd, and the high two bits of `left` are equal. For tagged integers, `right` is indeed always odd (although if this function is only intended to work for tagged integer comparisons rather than all comparisons, this should be documented), but the high two bits of `left` can sometimes differ. Concretely, this miscompiles `if x lsl 1 + 1 > 0 then ... else ...` (the optimised code does the wrong thing when `x = max_int`).

(Getting Cmm optimisations right in the presence of tagging is [really tricky](https://github.com/ocaml/ocaml/pull/863#issuecomment-282577836)).